paddle/platform/assert.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#define STRINGIFY(x) #x
+#define TOSTRING(x) STRINGIFY(x)
+
+#if defined(__APPLE__) && defined(__CUDA_ARCH__) && !defined(NDEBUG)
+#include <stdio.h>
+#define PADDLE_ASSERT(e)                                           \
+  do {                                                             \
+    if (!(e)) {                                                    \
+      printf("%s:%d Assertion `%s` failed.\n", __FILE__, __LINE__, \
+             TOSTRING(e));                                         \
+      asm("trap;");                                                \
```

Comment:
Sounds cool!

Also, after reading [this post](https://stackoverflow.com/questions/2080364/using-assert-within-kernel-invocation), I guess the current code is correct and no need to modify.

> CUDA now has a native assert function. Use assert(...). If its argument is zero, it will stop kernel execution and return an error. (or trigger a breakpoint if in CUDA debugging.)

> Make sure to include "assert.h". Also, this requires compute capability 2.x or higher, and is 
> not supported on MacOS. For more details see CUDA C Programming Guide, Section B.16.

> The programming guide also includes this example:

```c++
#include <assert.h>
__global__ void testAssert(void)
{
   int is_one = 1;
   int should_be_one = 0;
   // This will have no effect
   assert(is_one);
   // This will halt kernel execution
   assert(should_be_one);
}
int main(int argc, char* argv[])
{
   testAssert<<<1,1>>>();
   cudaDeviceSynchronize();
   return 0;
}
```